driver/src/main/java/org/neo4j/driver/internal/Bookmark.java

```diff
@@ -0,0 +1,172 @@
+{
+    private static final String BOOKMARK_KEY = "bookmark";
+    private static final String BOOKMARKS_KEY = "bookmarks";
+    private static final String BOOKMARK_PREFIX = "neo4j:bookmark:v1:tx";
+
+    private static final long UNKNOWN_BOOKMARK_VALUE = -1;
+
+    private static final Bookmark EMPTY = new Bookmark( Collections.<String>emptySet() );
+
+    private final Collection<String> values;
+    private final String maxValue;
+
+    private Bookmark( Collection<String> values )
+    {
+        this.values = values;
+        this.maxValue = maxBookmark( values );
+    }
+
+    public static Bookmark empty()
+    {
+        return EMPTY;
+    }
+
+    public static Bookmark from( String value )
+    {
+        if ( value == null )
+        {
+            return empty();
+        }
+        return from( singleton( value ) );
+    }
+
+    public static Bookmark from( Collection<String> values )
+    {
+        if ( values == null || values.isEmpty() )
+        {
+            return empty();
+        }
+        return new Bookmark( values );
+    }
+
+    public boolean isEmpty()
+    {
+        return values.isEmpty();
+    }
+
+    public String asString()
+    {
+        return maxValue;
+    }
+
+    public Map<String,Value> asParameters()
+    {
+        if ( isEmpty() )
+        {
+            return emptyMap();
+        }
+
+        // Driver sends {bookmark: "max", bookmarks: ["one", "two", "max"]} instead of simple
+        // {bookmarks: ["one", "two", "max"]} for backwards compatibility reasons. Old servers can only accept single
+        // bookmark that is why driver has to parse and compare given list of bookmarks. This functionality will
+        // eventually be removed.
+        Map<String,Value> parameters = new HashMap<>( 4 );
+        parameters.put( BOOKMARK_KEY, value( maxValue ) );
+        parameters.put( BOOKMARKS_KEY, value( values ) );
+        return parameters;
+    }
+
+    @Override
+    public boolean equals( Object o )
+    {
+        if ( this == o )
+        {
+            return true;
+        }
+        if ( o == null || getClass() != o.getClass() )
+        {
+            return false;
+        }
+        Bookmark bookmark = (Bookmark) o;
+        return Objects.equals( values, bookmark.values );
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash( values );
+    }
+
+    @Override
+    public String toString()
+    {
+        return "Bookmark{values=" + values + "}";
+    }
+
+    private static String maxBookmark( Collection<String> bookmarks )
+    {
+        if ( bookmarks == null || bookmarks.isEmpty() )
+        {
+            return null;
+        }
+
+        Iterator<String> iterator = bookmarks.iterator();
+
+        String maxBookmark = iterator.next();
+        long maxValue = bookmarkValue( maxBookmark );
+
+        while ( iterator.hasNext() )
+        {
+            String bookmark = iterator.next();
+            long value = bookmarkValue( bookmark );
```

Comment:
value might be null for the following case:
bookmarks = new String[]{aBookmark, null, null};